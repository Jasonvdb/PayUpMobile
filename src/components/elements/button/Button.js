import React, { Component } from "react";
import { Text, TouchableOpacity, Linking } from "react-native";
import LinearGradient from "react-native-linear-gradient";
import PropTypes from "prop-types";
import theme from "../../../config/theme";

// 1. Changed to a class based component
class Button extends Component {
  constructor(props) {
    super(props);
  }

  // 2. Custom function called onPress TouchableOpacity
  onPressHandler = () => {
    const { onPress, url } = this.props;
    if (url) {
      Linking.openURL(url);
    }
    onPress();
  };

  render() {
    const { children, variant, style } = this.props;

    let labelStyle = styles.label;
    let containerStyle = { ...styles.container, ...style };
    let gradientContainerProps = {
      start: { x: 0, y: 0 },
      end: { x: 0, y: 1 },
      colors: theme.defaultButtonGradients
    };

    switch (variant) {
      case "cta": {
        labelStyle = { ...labelStyle, ...styles.ctaLabel };
        containerStyle = { ...containerStyle, ...styles.ctaContainer };
        gradientContainerProps = {
          ...gradientContainerProps,
          colors: theme.ctaButtonGradients
        };
        break;
      }
      case "text": {
        labelStyle = { ...labelStyle, ...styles.textLabel };
        containerStyle = { ...containerStyle, ...styles.textContainer };
        gradientContainerProps = {
          colors: ["transparent"]
        };
        break;
      }
    }

    //TODO don't use TouchableOpacity for android

    return (
      <TouchableOpacity onPress={this.onPressHandler}>
        <LinearGradient {...gradientContainerProps} style={containerStyle}>
          <Text style={labelStyle}>{children}</Text>
        </LinearGradient>
      </TouchableOpacity>
    );
  }
}

Button.defaultProps = {
  variant: "default",
  style: {},
  onPress: () => {}
};

Button.propTypes = {
  onPress: PropTypes.func,
  url: PropTypes.string,
  children: PropTypes.string,
  variant: PropTypes.oneOf(["default", "cta", "text"]),
  style: PropTypes.object
};

const styles = {
  container: {
    alignSelf: "stretch",
    justifyContent: "center",
    backgroundColor: "transparent",
    borderRadius: 30,
    height: 55,
    color: "#fff"
  },
  label: {
    alignSelf: "center",
    color: "#FFF",
    fontSize: 18,
    fontWeight: "200",
    textTransform: "uppercase"
  },
  ctaContainer: {
    height: 55
  },
  ctaLabel: {
    fontSize: 22
  },
  textContainer: {
    height: 45
  },
  textLabel: {
    color: theme.defaultButtonLabelColor,
    fontSize: 18
  }
};

export default Button;
