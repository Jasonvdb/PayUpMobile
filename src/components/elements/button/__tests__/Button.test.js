import React from "react";
import { shallow, render } from "enzyme";
import Button from "../Button";

const mockOpenURL = jest.fn();

// Set openURL module function to jest.fn
jest.mock("Linking", () => ({
  openURL: mockOpenURL
}));

describe("button", () => {
  const origConsole = console.error;
  beforeEach(() => {
    console.error = message => {
      if (message.indexOf("PascalCase" > -1)) {
        return;
      }

      origConsole(message);
    };
  });
  afterEach(() => {
    console.error = origConsole;
  });

  describe("Rendering", () => {
    it("should match to snapshot - CTA", () => {
      const component = render(<Button variant={"cta"}>CTA button</Button>);
      expect(component).toMatchSnapshot("CTA button snapshot");
    });

    it("should match to snapshot - Default", () => {
      const component = render(<Button>Default button</Button>);
      expect(component).toMatchSnapshot("Default button snapshot");
    });

    it("should match to snapshot - Text", () => {
      const component = render(<Button variant={"text"}>Text button</Button>);
      expect(component).toMatchSnapshot("Default button snapshot");
    });
  });

  describe("interaction", () => {
    describe("onPressHandler", () => {
      const mockOnPress = jest.fn();
      let instance;

      beforeEach(() => {
        instance = shallow(
          <Button onPress={mockOnPress} url="https://www.test.com">
            Press me
          </Button>
        ).instance();
        jest.clearAllMocks();
      });

      it("should call onPress", () => {
        instance.onPressHandler();
        expect(mockOnPress).toHaveBeenCalled();
        expect(mockOnPress).toHaveBeenCalledTimes(1);
      });

      it("should call openURL if url is provided", () => {
        instance.onPressHandler();
        expect(mockOpenURL).toHaveBeenCalled();
        expect(mockOpenURL).toHaveBeenCalledTimes(1);
        expect(mockOpenURL).toHaveBeenCalledWith("https://www.test.com");
      });

      it("should not call openURL if url is not provided", () => {
        const innerInstance = shallow(
          <Button onPress={mockOnPress}>test label</Button>
        ).instance();
        innerInstance.onPressHandler();
        expect(mockOpenURL).not.toHaveBeenCalled();
      });

      it("should not call anything if neither url nor onPress or is not provided", () => {
        const innerInstance = shallow(<Button>test label</Button>).instance();
        innerInstance.onPressHandler();
        expect(mockOpenURL).not.toHaveBeenCalled();
      });
    });
  });
});
