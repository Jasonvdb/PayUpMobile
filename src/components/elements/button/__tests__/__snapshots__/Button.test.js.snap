// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button Rendering should match to snapshot - CTA: CTA button snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "opacity:1",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "colors": "4294465446,4294924376",
          "endpoint": "[object Object]",
          "startpoint": "[object Object]",
          "style": "align-self:stretch;justify-content:center;background-color:transparent;border-radius:30px;height:55px;color:#fff",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "style": "align-self:center;color:#FFF;font-size:22px;font-weight:200;text-transform:uppercase",
            },
            "children": Array [
              Object {
                "data": "CTA button",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "text",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "style": undefined,
            },
          },
        ],
        "name": "bvlineargradient",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "colors": undefined,
          "endpoint": undefined,
          "startpoint": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "colors": undefined,
          "endpoint": undefined,
          "startpoint": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "view",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`button Rendering should match to snapshot - Default: Default button snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "opacity:1",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "colors": "4294924376,4294924376",
          "endpoint": "[object Object]",
          "startpoint": "[object Object]",
          "style": "align-self:stretch;justify-content:center;background-color:transparent;border-radius:30px;height:55px;color:#fff",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "style": "align-self:center;color:#FFF;font-size:18px;font-weight:200;text-transform:uppercase",
            },
            "children": Array [
              Object {
                "data": "Default button",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "text",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "style": undefined,
            },
          },
        ],
        "name": "bvlineargradient",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "colors": undefined,
          "endpoint": undefined,
          "startpoint": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "colors": undefined,
          "endpoint": undefined,
          "startpoint": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "view",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`button Rendering should match to snapshot - Text: Default button snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "opacity:1",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "style": "align-self:stretch;justify-content:center;background-color:transparent;border-radius:30px;height:45px;color:#fff",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "style": "align-self:center;color:#666666;font-size:18px;font-weight:200;text-transform:uppercase",
            },
            "children": Array [
              Object {
                "data": "Text button",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "text",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "style": undefined,
            },
          },
        ],
        "name": "view",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "style": undefined,
        },
      },
    ],
    "name": "view",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
